File: Source/Samples/95_Periodic/code_aeditfield.cpp
// AEditField application source code
#include <Urho3D/Resource/ResourceCache.h>
#include <Urho3D/AUI/AEditField.h>
#include "PeriodicApp.h"

void PeriodicApp::setup_aeditfield( AWidget *layout )
{
    PODVector<AWidget*> dest;
    layout->SearchWidgetClass( "TBButton", dest );
    for (unsigned ii = 0; ii < dest.Size(); ii++)
        SubscribeToEvent(dest[ii], E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAeditfieldEvent ));

    AWidget *ed1 = layout->GetWidget("editfieldsingle");
    if ( ed1)
        SubscribeToEvent(ed1, E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAeditfieldEvent ));
    AWidget *ed2 = layout->GetWidget("editfieldmulti");
    if ( ed2)
        SubscribeToEvent(ed2, E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAeditfieldEvent ));
}

void PeriodicApp::HandleAeditfieldEvent(StringHash eventType, VariantMap& eventData)
{
    using namespace WidgetEvent;
    AWidget* widget = static_cast<AWidget*>(eventData[P_TARGET].GetPtr());
    if ( widget == NULL ) return;
    if (eventData[P_TYPE] == UI_EVENT_TYPE_CLICK)
    {
        if (widget->GetId() == "uieditfieldcode" )
        {
            AppLog( "AEditField support : " + widget->GetId() + " was pressed " );
            ViewCode ( "Sources/code_aeditfield.cpp", widget->GetParent() );
        }
        if (widget->GetId() ==  "uieditfieldlayout" )
        {
            AppLog( "AEditField support : " + widget->GetId() + " was pressed ");
            ViewCode ( "Scenes/layout_aeditfield.ui.txt", widget->GetParent() );
        }

        if (widget->GetId() ==  "editfieldadd" )
        {
            AppLog( "AEditField action : " + widget->GetId() + " was pressed ");
            AEditField *ef1 = static_cast<AEditField*>(widget->FindWidget("editfieldmulti"));
            if(ef1)
            {
                ResourceCache* cache = GetSubsystem<ResourceCache>();
                SharedPtr<File> filex = cache->GetFile("Scenes/layout_aeditfield.ui.txt");
                String textx = filex->ReadText();
                filex->Close();
                ef1->SetText(textx);
            }
        }
        if (widget->GetId() ==  "editfieldclr" )
        {
            AppLog( "AEditField action : " + widget->GetId() + " was pressed ");
            AEditField *ef2 = static_cast<AEditField*>(widget->FindWidget("editfieldmulti"));
            if(ef2)
                ef2->SetText("");
        }
    }
    else
    {
        if ( widget->GetId() ==  "editfieldsingle" )
        {
            AEditField *efx = static_cast<AEditField*>(widget);
            AppLog( "AEditField event : " + widget->GetId() + " text = `" + efx->GetText() + "` event type = " + EventReport(eventData[P_TYPE].GetInt()));
        }
        if ( widget->GetId() ==  "editfieldmulti" )
        {
            AEditField *efx = static_cast<AEditField*>(widget);
            AppLog( "AEditField event : " + widget->GetId() + " text = `" + efx->GetText() + "` event type = " + EventReport(eventData[P_TYPE].GetInt()));
        }
    }
}

File: bin/Data/AUI/Periodic/LuaScripts/code_aeditfield.lua
-- AEditField application source code

function setup_uieditfield( layout )
    local dest = layout:SearchAWidgetClass( "TBButton" )
    for i, awid in ipairs(dest) do
        SubscribeToEvent(awid, "WidgetEvent", "HandleUieditfieldEvent")
    end

    local ed1 = layout:GetWidget ("editfieldsingle")
    if ed1 ~= nil then
        local d1 = tolua.cast(ed1, "AWidget")
        SubscribeToEvent(d1, "WidgetEvent", "HandleUieditfieldEvent")
    end
    local ed2 = layout:GetWidget ("editfieldmulti")
    if ed2 ~= nil then
        local d2 = tolua.cast(ed2, "AWidget")
        SubscribeToEvent(d2, "WidgetEvent", "HandleUieditfieldEvent")
    end
end

function HandleUieditfieldEvent( eventType, eventData )
    local widget = eventData["Target"]:GetPtr("AWidget")
    if widget == nil then
        return;
    end
    if eventData["Type"]:GetInt() == UI_EVENT_TYPE_CLICK then
        if widget:GetId() == "uieditfieldcode" then
            AppLog( "AEditField support : " .. widget:GetId() .. " was pressed " )
            ViewCode ( "LuaScripts/code_aeditfield.lua", widget:GetParent())
        end
        if widget:GetId() ==  "uieditfieldlayout" then
            AppLog( "AEditField support : " .. widget:GetId() .. " was pressed ")
            ViewCode ( "Scenes/layout_aeditfield.ui.txt", widget:GetParent())
        end
        if widget:GetId() == "editfieldadd" then
            AppLog( "AEditField action : " .. widget:GetId() .. " was pressed ")
            local ef1 = widget:FindWidget("editfieldmulti")
            if ef1 ~= nil then
                local ef2 = tolua.cast(ef1, "AEditField")
                local filex = cache:GetFile("Scenes/layout_aeditfield.ui.txt")
                local textx = filex:ReadString()
                filex:Close()
                ef2:SetText(textx)
            end
        end
        if widget:GetId() ==  "editfieldclr" then
            AppLog( "AEditField action : " .. widget:GetId() .. " was pressed ")
            local ef1 = widget:FindWidget("editfieldmulti")
            if ef1 ~= nil then
                local ef2 = tolua.cast(ef1, "AEditField")
                ef2:SetText("")
            end
        end
    else
        if widget:GetId() == "editfieldsingle" then
           AppLog( "AEditField event : " .. widget:GetId() .. " text = `" .. widget:GetText())
        end
        if widget:GetId() ==  "editfieldmulti" then
            AppLog( "AEditField event : " .. widget:GetId() .. " text = `" .. widget:GetText())
        end
    end
end

